packages/gatsby/src/utils/start-server.ts

```diff
@@ -280,15 +248,183 @@ export async function startServer(
+        // For older, future, or otherwise unexpected stacks
+      } else {
+        splitLine = stackObject[0].split(`:`)
+      }
+      const splitLength = splitLine.length
+      filename = splitLine[splitLength - 3]
+      line = Number(splitLine[splitLength - 2])
+      row = Number(splitLine[splitLength - 1])
+    } catch (err) {
+      filename = ``
+      line = 0
+      row = 0
+    }
+    return {
+      filename: filename,
+      line: line,
+      row: row,
+    }
+  }
+  // Colors taken from Gatsby's design tokens
+  // https://github.com/gatsbyjs/gatsby/blob/d8acab3a135fa8250a0eb3a47c67300dde6eae32/packages/gatsby-design-tokens/src/colors.js#L185-L205
+  const colors = {
+    background: `fdfaf6`,
+    text: `452475`,
+    green: `137886`,
+    darkGreen: `006500`,
+    comment: `527713`,
+    keyword: `096fb3`,
+    yellow: `DB3A00`,
+  }
+
+  interface IParsedError {
+    filename: string
+    code: string
+    codeFrame: string
+    line: number
+    row: number
+    message: string
+    type: string
+    stack: [string]
+  }
+
+  const parseError = function (err): IParsedError {
+    const stack = err.stack ? err.stack : ``
+    const stackObject = stack.split(`\n`)
+    const position = getPosition(stackObject)
+    // Remove the `/lib/` added by webpack
+    const filename = path.join(
+      directory,
+      ...position.filename.split(path.sep).slice(2)
+    )
+    const code = fs.readFileSync(filename, `utf-8`)
+    const line = position.line
+    const row = position.row
+    ansiHTML.setColors({
+      reset: [colors.text, colors.background], // FOREGROUND-COLOR or [FOREGROUND-COLOR] or [, BACKGROUND-COLOR] or [FOREGROUND-COLOR, BACKGROUND-COLOR]
+      black: `aaa`, // String
+      red: colors.keyword,
+      green: colors.green,
+      yellow: colors.yellow,
+      blue: `eee`,
+      magenta: `fff`,
+      cyan: colors.darkGreen,
+      lightgrey: `888`,
+      darkgrey: colors.comment,
+    })
+    const codeFrame = ansiHTML(
+      codeFrameColumns(
+        code,
+        {
+          start: { line: line, column: row },
+        },
+        { forceColor: true }
+      )
+    )
+    const splitMessage = err.message ? err.message.split(`\n`) : [``]
+    const message = splitMessage[splitMessage.length - 1]
+    const type = err.type ? err.type : err.name
+    const data = {
+      filename: filename,
+      code,
+      codeFrame,
+      line: line,
+      row: row,
+      message: message,
+      type: type,
+      stack: stack,
+    }
+    return data
+  }
+
   // Render an HTML page and serve it.
-  app.use((_, res) => {
-    res.sendFile(directoryPath(`public/index.html`), err => {
-      if (err) {
-        res.status(500).end()
+  app.use(async (req, res, next) => {
+    const { pages } = store.getState()
+
+    if (!pages.has(req.path)) {
+      return next()
+    }
+
+    // Sleep until any work the server is doing has finished.
+    await new Promise(resolve => {
```

Comment:
@ascorbic this is what I meant

Reply:
@KyleAMathews OK, I get what you mean now. I don't think this is the right way to do it. As soon as you're watching a service and waiting for a specific state, it tells me that you should be a service inside the state machine. Luckily `app` is available in the state machine context, so you can [spawn a callback actor](https://xstate.js.org/docs/guides/actors.html#spawning-callbacks). Look at the `spawnWebpackListener` action for an example. You can pass that actor the `app` object and set up the handler in there. Then, when a request comes in you register an `onReceive` handler to resolve your promise, and then dispatch something like a `REQUEST_RECEIVED` event to the parent, and then in the parent it can either immediately respond with a `SEND_RESPONSES` reply if it's in `waiting`, or otherwise send one when you enter the waiting state. Happy to have a call to discuss this more, as it's a bit hard to explain here

Reply:
yeah this make sense – I'm starting to waver a bit on doing this though as we'd want to tightly limit how long the handler waits for changes e.g. 2-3 seconds — it's better at some point to show stale data than to keep waiting as the update would get pushed to the client once it's done.

Hmm though yeah I could still do this in xstate — just set a timeout and send the event early.

Lemme give this a go today on refactoring & I'll ping you if I need help.